File: src/main/java/java_game/Snake.java
package java_game;

import java.util.ArrayList;
import java.util.List;

public class Snake {

    private List<Pair> body = new ArrayList<>();
    private Direction direction;
    private boolean eat = false;

    public Snake(int x, int y, int length, Direction direction) {
        this.direction = direction;

        for (int i = 0; i < length; i++) {
            int bodyX = x;
            int bodyY = y;

            switch (direction){
                case DOWN:
                    bodyY = bodyY - i;
                    break;
                case UP:
                    bodyY = bodyY + i;
                    break;
                case RIGHT:
                    bodyX = bodyX - i;
                    break;
                case LEFT:
                    bodyX = bodyX + i;
                    break;
            }

            body.add(new Pair(bodyX, bodyY));
        }
    }

    public List<Pair> getBody() {
        return body;
    }

    public Direction getDirection() {
        return direction;
    }

    public void makeStep(Direction direction) {
        this.direction = direction;

        Pair newBodyPart = null;

        if (eat) {
            newBodyPart = new Pair(getTail().getX(), getTail().getY());
        }

        for (int i = body.size() - 1; i > 0; i--) {
            body.get(i).setX(body.get(i - 1).getX());
            body.get(i).setY(body.get(i - 1).getY());
        }

        if (direction == Direction.LEFT) {
            getHead().setX(getHead().getX() - 1);
        }
        if (direction == Direction.RIGHT) {
            getHead().setX(getHead().getX() + 1);
        }
        if (direction == Direction.UP) {
            getHead().setY(getHead().getY() - 1);
        }
        if (direction == Direction.DOWN) {
            getHead().setY(getHead().getY() + 1);
        }

        if (eat) {
            body.add(newBodyPart);
            eat = false;
        }
    }

    public void makeStep() {
        makeStep(direction);
    }

    public Pair getHead(){
        return body.get(0);
    }

    public Pair getTail(){
        return body.get(body.size() - 1);
    }

    public void setEat(boolean eat) {
        this.eat = eat;
    }
}

File: src/main/kotlin/kotlin_game/Snake.kt
package kotlin_game

class Snake(
        private val x: Int,
        private val y: Int,
        private val length: Int,
        private var direction: Direction
) {

    private val body: MutableList<Pair> = ArrayList()
    var eat: Boolean = false

    init {
        for(i in 0 .. length) {
            body.add( when(direction) {
                        Direction.DOWN -> Pair(x, y - i)
                        Direction.UP -> Pair(x, y + i)
                        Direction.RIGHT -> Pair(x - i, y)
                        Direction.LEFT -> Pair(x + i, y)
            })
        }
    }

    fun makeStep(direction: Direction) {
        this.direction = direction

        val newBodyPart = if(eat) {
            getTail().copy()
        } else {
            null
        }

        for(i in body.size - 1 downTo 1) {
            body[i] = body[i - 1].copy()
        }

        when(direction) {
            Direction.LEFT -> getHead().x = getHead().x - 1
            Direction.RIGHT -> getHead().x = getHead().x + 1
            Direction.UP -> getHead().y = getHead().y - 1
            Direction.DOWN -> getHead().y = getHead().y + 1
        }

        newBodyPart?.let {
            body.add(it)
            eat = false
        }
    }

    fun makeStep() {
        makeStep(direction)
    }

    fun getBody() = body
    fun getDirection() = direction
    fun getHead() = body.first()
    fun getTail() = body.last()
}